torch/csrc/jit/autodiff.cpp

```diff
@@ -34,38 +45,233 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
+      Value * new_grad = addAndPutAfter(prev_grad, dx, dx->node());
+      grad_map[x] = new_grad;
+    } else {
+      grad_map[x] = dx;
+    }
+  };
 
-  std::unordered_map<Value*, Value*> grad_map; // x -> dx mapping
-  const auto get_grad = [&](Value* v) { return grad_map[v]; };
-  for (auto output : graph->outputs())
-    grad_map[output] = graph->addInput()->setType(output->typeOption());
+  auto outputs = graph.outputs();
+  for (std::size_t i = 0, num_outputs = outputs.size(); i < num_outputs; ++i) {
+    Value * output = outputs[i];
+    Value * output_grad = graph.addInput()->setType(output->typeOption());
+    set_grad(output, output_grad);
+    grad_desc.df_input_vjps.push_back(i);
+  }
 
-  for (auto it = graph->rbegin(), end = graph->rend(); it != end; ++it) {
+  for (auto it = graph.rbegin(), end = graph.rend(); it != end; ++it) {
     Node *node = *it;
     auto inputs = node->inputs();
     value_list grad_inputs = gradientForNode(node, fmap(node->outputs(), get_grad));
     JIT_ASSERT(grad_inputs.size() == node->inputs().size());
     for (std::size_t i = 0, num_inputs = grad_inputs.size(); i < num_inputs; ++i) {
-      if (Value * prev_grad = grad_map[inputs[i]]) {
-        Node *new_grad_node = graph->create(kadd, {prev_grad, grad_inputs[i]})
-                                   ->t_(kalpha, at::Scalar(1).toTensor());
-        new_grad_node->insertAfter(grad_inputs[i]->node());
-        Value *new_grad = new_grad_node->output();
-        new_grad->setType(prev_grad->typeOption());
-        grad_map[inputs[i]] = new_grad;
-      } else {
-        grad_map[inputs[i]] = grad_inputs[i];
-      }
+      set_grad(inputs[i], grad_inputs[i]);
     }
   }
 
-  for (auto input : graph->inputs()) {
+  auto inputs = graph.inputs();
+  for (std::size_t i = 0, num_inputs = inputs.size(); i < num_inputs; ++i) {
+    Value * input = inputs[i];
     if (input->stage() > 0) break;
-    graph->registerOutput(grad_map.at(input));
+    graph.registerOutput(get_grad(input));
+    grad_desc.df_output_vjps.push_back(i);
+  }
+
+  return grad_map;
+}
+
+// This function will take the graph and return a new one that:
+//   - contains all nodes of graph that have given stage
+//   - there will be an input corresponding to each input of the inputs array
+//   - values corresponding to outputs will be returned from the new graph
+// It requires that values contained in inputs are sufficient to be able to
+// compute all values in a given stage. An exception will be thrown if this is
+// not the case.
+static std::shared_ptr<Graph> splitOffStage(
+        Graph& graph,
+        std::size_t stage,
+        ArrayRef<Value*> inputs,
+        ArrayRef<Value*> outputs) {
+  auto graph_clone = std::make_shared<Graph>();
+
+  value_map val_map; // values in graph -> values in graph_clone
+  const auto lookup_val = [&](Value *v) { return val_map.at(v); };
+
+  for (Value *input : inputs)
+    val_map[input] = graph_clone->addInput()->setType(input->typeOption());
```

Comment:
Opportunity for additional safety here: assert that the stage of input is <= stage.